pulsar-client/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java

```diff
@@ -80,6 +82,38 @@ public void testClientCnxTimeout() throws Exception {
         eventLoop.shutdownGracefully();
     }
 
+    @Test
+    public void testPendingLookupRequestSemaphore() throws Exception {
+        EventLoopGroup eventLoop = EventLoopUtil.newEventLoopGroup(1, false, new DefaultThreadFactory("testClientCnxTimeout"));
+        ClientConfigurationData conf = new ClientConfigurationData();
+        conf.setOperationTimeoutMs(10_000);
+        conf.setKeepAliveIntervalSeconds(0);
+        ClientCnx cnx = new ClientCnx(conf, eventLoop);
+
+        ChannelHandlerContext ctx = mock(ChannelHandlerContext.class);
+        Channel channel = mock(Channel.class);
+        when(ctx.channel()).thenReturn(channel);
+        ChannelFuture listenerFuture = mock(ChannelFuture.class);
+        when(listenerFuture.addListener(any())).thenReturn(listenerFuture);
+        when(ctx.writeAndFlush(any())).thenReturn(listenerFuture);
+        cnx.channelActive(ctx);
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        new Thread(() -> {
+            try {
+                cnx.newLookup(null, 123).get();
+            } catch (Exception e) {
+                // ignore exception
+            }
+            countDownLatch.countDown();
+        }).start();
+        cnx.channelInactive(ctx);
+        countDownLatch.await();
+        // wait for subsequent calls over
+        Thread.sleep(1_000);
```

Comment:
Looks like we can use Awaitibility instead.

Reply:
fixed